### PR TITLE
Fix FailbackClusterInvoker one risk of memory leak #2425

### PR DESCRIPTION
**Pull request by wangweiufofly**
1.limit the size of the map,default is 1000. can config by key failcapacity
2.retry period,default is 100.can config by key retries.
3.shutdown the timer when destroyed

The issues:
https://github.com/apache/incubator-dubbo/issues/2425

### CODE REVIEW

File: dubbo-config/dubbo-config-spring/src/main/resources/META-INF/dubbo.xsd [hunk 1]
```diff
@@ -25,6 +25,11 @@
                 <xsd:documentation><![CDATA[ The method retry times. ]]></xsd:documentation>
             </xsd:annotation>
         </xsd:attribute>
+        <xsd:attribute name="failcapacity" type="xsd:string">
```
[the remark this answers is not in the document]
Comment (wangweiufofly) [reply dubbo-config/dubbo-config-spring/src/main/resources/META-INF/dubbo.xsd]: Good idea. I've changed it

File: dubbo-common/src/main/java/org/apache/dubbo/common/Constants.java [hunk 2]
```diff
@@ -139,6 +139,10 @@
 
     public static final int DEFAULT_RETRIES = 2;
 
+    public static final int DEFAULT_FAIL_CAPACITY_SIZE = 1000;
```
[the remark this answers is not in the document]
Comment (wangweiufofly) [reply dubbo-common/src/main/java/org/apache/dubbo/common/Constants.java]: Good idea. I've changed it.

File: dubbo-common/src/main/java/org/apache/dubbo/common/Constants.java [hunk 3]
```diff
@@ -139,6 +139,10 @@
...
+
+    public static final int DEFAULT_FAIL_RETRY_SIZE = 100;
```
[the remark this answers is not in the document]
Comment (wangweiufofly) [reply dubbo-common/src/main/java/org/apache/dubbo/common/Constants.java]: Good idea. I've changed it.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/FailbackClusterInvoker.java [hunk 4]
```diff
@@ -16,90 +16,75 @@
...
-     * Use {@link NamedInternalThreadFactory} to produce {@link org.apache.dubbo.common.threadlocal.InternalThread}
-     * which with the use of {@link org.apache.dubbo.common.threadlocal.InternalThreadLocal} in {@link RpcContext}.
-     */
-    private final ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(2,
-            new NamedInternalThreadFactory("failback-cluster-timer", true));
+    private final int retries;
 
-    private final ConcurrentMap<Invocation, AbstractClusterInvoker<?>> failed = new ConcurrentHashMap<>();
-    private volatile ScheduledFuture<?> retryFuture;
+    private final int failCapacitySize;
```
[the remark this answers is not in the document]
Comment (wangweiufofly) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/FailbackClusterInvoker.java]: Good idea. I've changed it.

File: dubbo-common/src/main/java/org/apache/dubbo/common/Constants.java [hunk 5]
```diff
@@ -272,6 +276,8 @@
 
     public static final String RETRIES_KEY = "retries";
 
+    public static final String FAIL_CAPACITY_KEY = "failcapacity";
```
[the remark this answers is not in the document]
Comment (wangweiufofly) [reply dubbo-common/src/main/java/org/apache/dubbo/common/Constants.java]: Good idea. I've changed it.

**Comment by carryxyh** [comment]
@beiwei30 @chickenlj 
Would u pls have a look?

**Comment by wangweiufofly** [comment]
> Hi, @wangweiufofly
> I think we should add a ut for this feature.
> We should test the size of failback tasks and the retry times. :)

test case completed

File: dubbo-cluster/src/test/java/org/apache/dubbo/rpc/cluster/support/FailbackClusterInvokerTest.java [hunk 6]
```diff
@@ -118,15 +120,21 @@ public void testNoInvoke() {
...
         FailbackClusterInvoker<FailbackClusterInvokerTest> invoker = new FailbackClusterInvoker<FailbackClusterInvokerTest>(
                 dic);
         invoker.invoke(invocation);
+        invoker.invoke(invocation);
+        invoker.invoke(invocation);
         Assert.assertNull(RpcContext.getContext().getInvoker());
-        invoker.retryFailed();// when retry the invoker which get from failed map already is not the mocked invoker,so
+//        invoker.retryFailed();// when retry the invoker which get from failed map already is not the mocked invoker,so
+        //Ensure that the main thread is online
+        CountDownLatch countDown = new CountDownLatch(1);
```
Comment (carryxyh) [inline dubbo-cluster/src/test/java/org/apache/dubbo/rpc/cluster/support/FailbackClusterInvokerTest.java]: I think u should assert that the logger contains the specific msg.

File: dubbo-cluster/src/test/java/org/apache/dubbo/rpc/cluster/support/FailbackClusterInvokerTest.java [hunk 7]
```diff
@@ -112,21 +119,34 @@ public void testNoInvoke() {
...
         LogUtil.start();
+        DubboAppender.clear();
         invoker.invoke(invocation);
         assertEquals(1, LogUtil.findMessage("Failback to invoke"));
         LogUtil.stop();
     }
 
     @Test()
-    public void testRetryFailed() {
+    public void testARetryFailed() throws Exception {
```
Comment (carryxyh) [inline dubbo-cluster/src/test/java/org/apache/dubbo/rpc/cluster/support/FailbackClusterInvokerTest.java]: Could u pls tell us why u rename the method and why u add the FixMethodOrder annotation?
:)

**Pull request merged**